fetchRebalancerData.js

```diff
@@ -9,68 +11,94 @@ const contractAbi = require('./UniswapNFTPositionManagerAbi.json');
+                    liquidityRebalanced: e.returnValues.liquidity
+                };
+            })
+        // append the list for current block range into the overall list
+        allValidTransactions = allValidTransactions.concat(validTransactionInRange);
+
+        // update currentFirstBlock for making the next set of queries
+        currentFirstBlock = currentToBlock + 1;
+    }
+
+    // get transaction data of each transaction hash obtained above 
+    let totalLiquidityRebalanced = 0;
+
+    transactionData = await Promise.all(allValidTransactions.map(async (transaction) => {
         const result = await web3.eth.getTransaction(transaction.transactionHash);
         return {
-            tokenId: transaction.tokenId,
-            liquidity: transaction.liquidity,
-            to: result.to,
-            from: result.from,
-            hash: result.hash
+            newTokenId: transaction.newTokenId,
+            oldTokenId: transaction.oldTokenId,
+            dsaAddr: result.to,
+            transactionHash: result.hash,
+            liquidityRebalanced: transaction.liquidityRebalanced
         }
     }));
 
-    // console.log(finalTransactions);
+    // keep only the transactions that were made
+    const instadappRebalanceTransactions = transactionData.filter(transaction => dsaAccounts.includes(transaction.dsaAddr.toLowerCase()));
 
-    const idappTransactions = finalTransactions.filter(transaction => transaction.to.toLowerCase() === instaContractAddress.toLowerCase())
-    console.log(idappTransactions);
-    
-    const users = new Set();
+    // create Sets to store unique users and tokenIds
     const tokenPools = new Set();
-    let totalLiquidityRebalanced = 0;
 
-    idappTransactions.forEach(transaction => {
-        tokenPools.add(transaction.tokenId);
-        users.add(transaction.from);
-        totalLiquidityRebalanced += Number(transaction.liquidity);
-    })
+    instadappRebalanceTransactions.forEach(transaction => {
+        // add tokenIds and users of each transaction into respective sets
+        tokenPools.add(transaction.oldTokenId);
+        tokenPools.add(transaction.newTokenId);
+        totalLiquidityRebalanced += Number(transaction.liquidityRebalanced);
+    });
+    // console.log(convertLiquidityToTokens);
+    const transactions = await convertLiquidityToTokens(instadappRebalanceTransactions);
 
-    console.log("Users: ", users);
-    console.log("No. of unique users: ", users.size);
+    // console.log("List of all Rebalance Transactions on Instadapp: ", instadappRebalanceTransactions);
+    console.log("List of all Rebalance Transactions with Token info: ", transactions);
+    console.log("Total no. of times the strategy got used: ", instadappRebalanceTransactions.length);
     console.log("Token Pools: ", tokenPools);
-    console.log("Total Liquidity rebalanced: ", totalLiquidityRebalanced)
+    console.log("Total Liquidity rebalanced: ", totalLiquidityRebalanced);
```

Comment:
Can we get this in USD?

Reply:
Done. Set `totalLiquidityRebalancedInUsd` to equal sum of `token0Usd + token1Usd` for all transactions.